File: packages/shared-schema/src/opspilot_schema/change_impact.py
from __future__ import annotations
from typing import Literal
from pydantic import BaseModel


class ChangeImpactRequest(BaseModel):
    change_type: str
    target_type: str
    target_id: str
    requested_action: str
    environment: str
    change_window: str | None = None


class ImpactedObject(BaseModel):
    object_type: str
    object_id: str
    object_name: str
    impact_type: str
    severity: str


class DependencyNode(BaseModel):
    id: str
    name: str
    type: str
    children: list[DependencyNode] = []


class ChangeHypothesis(BaseModel):
    id: str
    summary: str
    confidence: float
    support_evidence_refs: list[str] = []
    counter_evidence_refs: list[str] = []
    missing_evidence: list[str] = []
    status: Literal["candidate", "confirmed", "probable", "refuted", "inconclusive"] = "candidate"


class ChangeImpactResult(BaseModel):
    analysis_id: str
    target: dict
    action: str
    risk_score: int
    risk_level: Literal["critical", "high", "medium", "low"]
    impacted_objects: list[ImpactedObject] = []
    checks_required: list[str] = []
    rollback_plan: list[str] = []
    approval_suggestion: Literal["required", "recommended", "not_required"]
    dependency_graph: list[DependencyNode] = []
    evidence_sufficiency: dict | None = None
    conclusion_status: Literal["confirmed", "probable", "insufficient_evidence", "contradicted"] | None = None
    counter_evidence_result: dict | None = None
    hypotheses: list[ChangeHypothesis] = []

File: packages/shared-types/src/change-impact.ts
export type RiskLevel = "critical" | "high" | "medium" | "low";

export interface ChangeImpactRequest {
  change_type: string;
  target_type: string;
  target_id: string;
  requested_action: string;
  environment: string;
  change_window?: string;
}

export interface ImpactedObject {
  object_type: string;
  object_id: string;
  object_name: string;
  impact_type: string;
  severity: string;
}

export interface ChangeImpactResult {
  analysis_id: string;
  target: {
    target_type: string;
    target_id: string;
    environment: string;
    type?: string;
    id?: string;
    name?: string;
  };
  action: string;
  risk_score: number;
  risk_level: RiskLevel;
  impacted_objects: ImpactedObject[];
  checks_required: string[];
  rollback_plan: string[];
  approval_suggestion: "required" | "recommended" | "not_required";
  dependency_graph: DependencyNode[];
  evidence_sufficiency?: {
    required_evidence_types: string[];
    present_evidence_types: string[];
    missing_critical_evidence: string[];
    sufficiency_score: number;
    freshness_score: number;
  };
  conclusion_status?: "confirmed" | "probable" | "insufficient_evidence" | "contradicted";
  counter_evidence_result?: {
    status: "refuted" | "not_refuted" | "inconclusive";
    checked_hypothesis_id?: string | null;
    summary: string;
    evidence_refs: string[];
  };
  hypotheses?: Array<{
    id: string;
    summary: string;
    confidence: number;
    support_evidence_refs: string[];
    counter_evidence_refs: string[];
    missing_evidence: string[];
    status: "candidate" | "confirmed" | "probable" | "refuted" | "inconclusive";
  }>;
}

export interface DependencyNode {
  id: string;
  name: string;
  type: string;
  children: DependencyNode[];
}
